Troubleshooting: add 'Editors Values are not Saved' article (#5107)
DevExpress/devextreme-documentation: concepts/80 Troubleshooting/05 DataGrid and TreeList Issues/00 Editors Values are not Saved.md

> If an editor does not pass its value to the DataGrid/TreeList and the component's DataSource when users enter/choose a value, check if any of the following cases applies to your implementation.

Reviewer: **First**. Multiple entities delimited by slash characters. In general, I advise against using them because they definitely do not add unambiguity to your wordings. Of what do I speak?

You can read "DataGrid/TreeList" in three ways, actually:

1. DataGrid **or** TreeList
2. DataGrid **and** TreeList
3. DataGrid **and/or** TreeList

When it's acceptable to use slash characters in a sentence? For example, when you need to cram two or more scenarios into one sentence for one reason or another:

> Use this interface to save/load data to/from a file.

This is a bit artificial example but it illustrates the concept. In your scenario, the slash character resolves into OR (the first interpretation, see above). Therefore, we should use OR instead of slash characters.

**Second**. An indirect synonym. First you say "the DataGrid/TreeList". Then refer to them/it as "the component". Yes, one probably knows that DataGrd and TreeList are components. Still, it's better to name them components before addressing them as such.

```suggestion
If an editor does not pass its value to the DataGrid or TreeList component and its DataSource when users enter or select a value, check if any of the following cases applies to your implementation.
```
P.S. What is "DataSource"? Is it a property? If yes, it probably should have been a link to the corresponding topic.

> ### A column's dataField is empty

Reviewer: You do not capitalize topic section headings, don't you? If you don't in all similar topics throughout your documentation repository -- then everything is fine. Otherwise, you should capitalize them properly.

> Built-in column editors automatically write their values to the data row field specified in the column.[dataField](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/#dataField) property. Specify this property if you want to write an editor's value to a specific field.

Reviewer: I'm afraid that the word "specific" does not add anything of value to this paragraph, mainly because if one leaves the mentioned property unspecified, editor values will go nowhere.

```suggestion
Built-in column editors automatically write their values to the data row field specified in the column.[dataField](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/#dataField) property. Specify this property if you want to write an editor's value to a field.
```

        <DxDataGrid ... >
            <DxColumn ...
                :set-cell-value="setCellValue"
            />
        </DxDataGrid>
    </template>

    <script>
        import { DxDataGrid, DxColumn } from 'devextreme-vue/data-grid';
        
        export default {
            components: {
                DxDataGrid,
                DxColumn
            },
            // ...
            methods: {
                setCellValue(newData, value, currentRowData) {
                    newData.anyField = value;
                }
            }
        };
    </script>

##### React

    <!-- tab: App.js -->
    import React from 'react';
    import DataGrid, { Column } from 'devextreme-react/data-grid';

    const setCellValue = (newData, value, currentRowData) => {
        newData.anyField = value;
    }

    function App() {
        render (
            <React.Fragment>
                <DataGrid ... >
                    <Column ...
                        setCellValue={setCellValue}
                    />
                </DataGrid>
            </React.Fragment>
        );
    }

    export default App;

---

### You use a form item's template to declare an editor

> The [form](/Documentation/Guide/UI_Components/DataGrid/Editing/#User_Interaction/Form_Mode) and [popup](/Documentation/Guide/UI_Components/DataGrid/Editing/#User_Interaction/Popup_Mode) edit modes use the built-in Form component. Do not use simpleItem.[template](/Documentation/ApiReference/UI_Components/dxForm/Item_Types/SimpleItem/#template) to replace default editors in DataGrid or TreeList, use column.[editCellTemplate](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/#editCellTemplate) instead.

Reviewer: > **The** `form` and `popup` edit modes

Every time you enumerate named entities joined with AND, you can omit a definite article, according to the new guideline on articles.

The rest looks good to me here.

```suggestion
[Form](/Documentation/Guide/UI_Components/DataGrid/Editing/#User_Interaction/Form_Mode) and [popup](/Documentation/Guide/UI_Components/DataGrid/Editing/#User_Interaction/Popup_Mode) edit modes use the built-in Form component. Do not use simpleItem.[template](/Documentation/ApiReference/UI_Components/dxForm/Item_Types/SimpleItem/#template) to replace default editors in DataGrid or TreeList, use column.[editCellTemplate](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/#editCellTemplate) instead.
```

### You implement an editor in the editCellTemplate body

> Call the **e.setValue** method available in the [template's argument](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/#editCellTemplate). This method notifies the DataGrid/TreeList that the value of a custom editor is changed. If you use a DevExtreme editor in this template, call **e.setValue** inside the [onValueChanged](/Documentation/ApiReference/UI_Components/dxTextBox/Configuration/#onValueChanged) event handler of this editor.

Reviewer: ```suggestion
Call the **e.setValue** method available in the [template's argument](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/#editCellTemplate). This method notifies the DataGrid or TreeList component that the value of a custom editor has changed. If you use a DevExtreme editor in this template, call **e.setValue** inside the [onValueChanged](/Documentation/ApiReference/UI_Components/dxTextBox/Configuration/#onValueChanged) event handler of this editor.
```

#include common-demobutton-named with {
    url: "https://js.devexpress.com/Demos/WidgetsGallery/Demo/DataGrid/CustomEditors/",
    name: "Customize Editors in DataGrid"
}

### You implement an editor in the cellTemplate body

> DataGrid/TreeList uses column.[cellTemplate](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/#cellTemplate) to display a cell value, not for editing. Use column.[editCellTemplate](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/#editCellTemplate) to place your custom editor in cells for editing. Refer to the previous paragraph for more information.

Reviewer: ```suggestion
DataGrid or TreeList uses column.[cellTemplate](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/#cellTemplate) only to display a cell value. To place your custom editor into cells to allow users to edit them, use column.[editCellTemplate](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/#editCellTemplate). Refer to the previous topic section for more information.
```

> If you want to always display editors in a column, enable the column.[showEditorAlways](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/#showEditorAlways) option. DataGrid displays an editor (or **editCellTemplate** if exists) and ignores **cellTemplate**.

Reviewer: ```suggestion
If you want to always display editors in a column, enable the column.[showEditorAlways](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/#showEditorAlways) option. DataGrid displays an editor (or **editCellTemplate**, if it exists) and ignores **cellTemplate**.
```

> If you want to always display editors in a column, enable the column.[showEditorAlways](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/#showEditorAlways) option. DataGrid displays an editor (or **editCellTemplate**, if it exists) and ignores **cellTemplate**.

Reviewer 2: ```suggestion
If you want to always display editors in a column, enable the column.[showEditorAlways](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/#showEditorAlways) option. DataGrid displays an editor (or **editCellTemplate** if it exists) and ignores **cellTemplate**.
```

                        const defaultValueChangeHandler = e.editorOptions.onValueChanged;
                        e.editorOptions.onValueChanged = function (args) {
                            e.setValue(args.value);
                            // or
                            defaultValueChangeHandler(args);
                        }
                    }
                }
            }
        };
    </script>

##### React

    <!-- tab: App.js -->
    import React from 'react';
    import DataGrid from 'devextreme-react/data-grid';

    onEditorPreparing = (e) => {
        if (e.dataField === 'requiredDataField' && e.parentType === 'dataRow') {
            const defaultValueChangeHandler = e.editorOptions.onValueChanged;
            e.editorOptions.onValueChanged = function (args) {
                e.setValue(args.value);
                // or
                defaultValueChangeHandler(args);
            }
        }
    }

    function App() {
        render (
            <React.Fragment>
                <DataGrid ... 
                    onEditorPreparing={onEditorPreparing}
                />
            </React.Fragment>
        );
    }

    export default App;

---

> #####See Also#####

Reviewer: Unlike other topic section headings, this one looks capitalized. This does not look consistent to me. However, if you do the same throughout all other such topics in the repository -- then everything is fine.

Just want to make sure that everything is fine -- nothing else.
